### PR TITLE
[GH-917] - Deathmatch mode

### LINKED ISSUE #917: [FEATURE] Create Deathmatch gamemode
We want to create another game mode

Create a new game mode that will go as follows:

- There's no zone that kills the player
- Players re-spawn some time after dying
- The match will end after a fixed time and the winner will be the player with most kills

#### Things to Change/Add

- **Game Mode Configuration**  
  Add the game mode to the game configuration (passed through matchmaking) to differentiate between *Battle Royale* and *Deathmatch* modes.

- **Game Initialization Parameters (for deathmatch)**  
  When the game is created, ensure the following parameters are set:
    - The zone is always disabled
    - Add a configurable game duration timer
    - Set configurable respawn seconds

- **End Game Check**  
  Keep `:end_game_check` as it currently is in *Battle Royale* mode, we won't use it for Deathmatch.  
  - Add an `if` statement to distinguish between *Battle Royale* :end_game_check and *Deathmatch*.
  - In *Deathmatch*, send a message to the game updater process to end the game (and retrieve the winner) after a configurable time.

- **Kill Tracking**  
  Track kills — ***already implemented***

- **Respawn Queue**  
  Add a pipe at the end of each tick update to send every dead player to a respawn queue (it's not really a queue, just a function that will be triggered after a fixed time), all of them will be respawned after a configurable time.

- **Respawn Implementation**  
  - Use `send after` to trigger respawn after a configurable time.
  - Respawn will restore the player to their initial health and position. (The position could be random too - 2nd iteration)

#### Second Iteration

- **Tiebreaker Algorithm**  
  Add an algorithm to resolve ties. -> (amount of deaths, random).

- **Invincibility After Respawn**  
  Include configurable invincibility seconds following respawn.

- **Game Duration to Frontend**  
  Send the current time until game finishes information to the frontend.

- **Podium Implementation**  
  Implement a podium to display player standings at the end of the game (top N kills).

### PR DESCRIPTION
**Pull request by tvillegas98**
> [!Warning]
> Merge alongside https://github.com/lambdaclass/champions_of_mirra/pull/2230
## Motivation

Add a new game mode based on the deathmatch concept. All against all mode where the player with most kills wins the match.

Closes #917

## Summary of changes

- Added a new matchmaking for deathmatch mode
- Added a respawn feature only enabled for deathmatch mode
- Added a new win condition only enabled for deathmatch mode
- Send the gamemode to the client

## How to test it?

- Open the browser client and play a game in the deathmatch mode. If you want, you can print the winner in `apps/arena/lib/arena/game_updater.ex:337` and check that it is correct

## Checklist
- [x] Tested the changes locally.
- [x] Reviewed the changes on GitHub, line by line.
- [ ] This change requires new documentation.
  - [ ] Documentation has been added/updated.

### CODE REVIEW

File: apps/arena/lib/arena/game/player.ex [hunk 1]
```diff
@@ -451,6 +451,14 @@ defmodule Arena.Game.Player do
     end)
   end
 
+  def respawn_player(player, position) do
+    aditional_info = player.aditional_info |> Map.put(:health, player.aditional_info.max_health)
```
Comment (Nico-Sanchez) [inline apps/arena/lib/arena/game/player.ex:455]: Shouldn't we use base_health?

File: apps/arena/lib/arena/game_updater.ex [hunk 2]
```diff
@@ -317,6 +325,31 @@ defmodule Arena.GameUpdater do
...
+  def handle_info(:deathmatch_end_game_check, state) do
+    players =
+      state.game_state.players
+      |> Enum.map(fn {player_id, player} ->
+        %{kills: kills} = GameTracker.get_player_result(player_id)
+        {player_id, player, kills}
+      end)
+      |> Enum.sort_by(fn {_player_id, _player, kills} -> kills end, :desc)
+
+    {winner_id, winner, _kills} = Enum.at(players, 1)
```
Comment (Nico-Sanchez) [inline apps/arena/lib/arena/game_updater.ex:337]: Shouldn't this be zero?
```suggestion
    {winner_id, winner, _kills} = Enum.at(players, 0)
```

File: apps/arena/lib/arena/game_updater.ex [hunk 3]
```diff
@@ -1873,6 +1911,52 @@ defmodule Arena.GameUpdater do
     end
   end
 
+  defp add_players_to_respawn_queue(game_state, %{game: %{game_mode: :DEATHMATCH}} = game_config) do
+    death_players =
+      game_state.players
+      |> Enum.filter(fn {_player_id, player} ->
+        not Player.alive?(player)
```
Comment (Nico-Sanchez) [inline apps/arena/lib/arena/game_updater.ex:1918]: couldn't we just check here if the player is not in the respawn queue already? That way we avoid the 'if' below and another iteration over the dead players.

File: apps/arena/lib/arena/matchmaking/deathmatch_mode.ex [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+  alias Ecto.UUID
+
+  use GenServer
+
+  # 3 Mins
+  @match_duration 180_000
+  @respawn_time 5000
+
+  # Time to wait to start game with any amount of clients
+  @start_timeout_ms 4_000
```
Comment (Nico-Sanchez) [inline apps/arena/lib/arena/matchmaking/deathmatch_mode.ex:13]: Create an issue and link it here to add these params to the Configurator.

File: apps/serialization/messages.proto [hunk 5]
```diff
@@ -69,10 +69,18 @@ message Configuration {
   ClientConfig client_config = 4;
 }
 
+enum GameMode{
+  BATTLE = 0;
+  DEATHMATCH = 1;
+  PAIR = 2;
+  QUICK_GAME = 3;
+}
```
Comment (Nico-Sanchez) [inline apps/serialization/messages.proto:77]: Remember to update arena's version. We do that everytime we touch the proto file (and other cases).

**Review by Nico-Sanchez — CHANGES_REQUESTED** [review]

File: apps/arena/lib/arena/game/player.ex (hunk 1 above)
Comment (tvillegas98) [reply apps/arena/lib/arena/game/player.ex:455]: Whoops, you're right

File: apps/arena/lib/arena/game_updater.ex (hunk 3 above)
Comment (tvillegas98) [reply apps/arena/lib/arena/game_updater.ex:1918]: We can just remove the first iteration and move everything to the second iteration

File: apps/serialization/messages.proto (hunk 5 above)
Comment (tvillegas98) [reply apps/serialization/messages.proto:77]: Sure!

File: apps/arena/lib/arena/matchmaking/deathmatch_mode.ex (hunk 4 above)
Comment (tvillegas98) [reply apps/arena/lib/arena/matchmaking/deathmatch_mode.ex:13]: Done in #985

**Comment by tvillegas98** [comment]
all changes were done in [chore: requested changes](https://github.com/lambdaclass/mirra_backend/pull/980/commits/ab95ea6e0bda80afa12f7e64d70c99898f529827)

**Review by Nico-Sanchez — DISMISSED** [review]

**Review by Nico-Sanchez — DISMISSED** [review]

**Review by Nico-Sanchez — APPROVED** [review]

File: apps/serialization/messages.proto
Comment (manucamejo) [inline apps/serialization/messages.proto:1]: Since we are changing protobuf, we should upgrade the mix version, right?

**Review by manucamejo — APPROVED** [review]

**Review by Nico-Sanchez — APPROVED** [review]

**Pull request merged**